CleverRaven/Cataclysm-DDA
# Switch crafter in crafting menu to and from an ally

#### Summary
Interface "Switch crafter in crafting menu to and from an ally"


### Purpose of change
Resolves #66163
![image](https://github.com/CleverRaven/Cataclysm-DDA/assets/13402666/eefa2e34-f9ad-4612-8d5c-35feffc01653)




### Describe the solution
- [x] Depends on #68666
- [x] Crafting group*.
- [x] Menu for switching crafters of crafting group. Listing all allies. Including those who cannot craft right now with an explanation of why it is so.
- [x] REMOVE TEMPORARY: Cycle through crafters with keybinding - includes those who should not be selectable (asleep, crafting, ...).
- [x] REMOVE TEMPORARY: List members of the crafting group at the top of detail.
- [x] List the crafter at the top of detail.
- [x] Show all recipes the crafting group knows.
- [x] Allow crafting only recipes you should be able to (like before - so don't introduce a bug).
- [x] Mark recipes (in the left column) beyond crafters' understanding. With light red (unused so far). Categories with magenta (unused so far), but I wanted purple (doesn't exist). ~With a new colour? Or strike them? - don't know how Suggestions welcome!~
- [x] In recipe detail, if a recipe is beyond understanding, colour skill requirement with red (eg. fabrication `<red>4/8</red>`). Also, add the text: `<color_red>Cannot be crafted because the crafter is not skilled enough to understand the recipe.</color>`.
- [ ] The above marks secondary skills as red too, should revert to yellow?
- [x] fix: NPC cannot be told a recipe they don't know by the player - this was in crafting from a conversation with NPC, `iuse.cpp`, and `iexamine.cpp`. Solved by separating `crafting_group()` (allies nearby) from `crafting_helpers()` (willing, awake, etc. who will help with batch crafting etc.)
- [x] fix: On non-memorized recipes, if it is both written in a book and known by an ally, only `Written in` was shown, now `Known by` is also shown.
- [x] Improve some text.
- [x] Characters provide crafting recipes and proficiencies even when sleeping (preventing optimal non-fun gameplay of waking everyone up before crafting and letting them fall asleep immediately after choosing a recipe). They still don't learn if asleep.
- [x] Don't duplicate code if not necessary (get_available_recipes, has_recipe).
- [x] fix: `Searched+` tab overlapped
- [x] fix: "Cannot be crafted because the same item is needed for multiple components" triggered for NPCs that cannot craft or practice this.

*__Crafting group__: on each opening of the crafting menu (be it with keybinding `&` or through conversation with allied NPC), the crafter finds nearby allies forming a `crafting_group` of which they are a part. The group shares their knowledge and you can switch between crafters, the default crafter is who it was before this PR. Recipes of all participants are shown, so switching crafters is more seamless. Recipes beyond the crafters' knowledge are marked as uncraftable.



### Describe alternatives you've considered



The default keybinding for choosing a crafter is `c`, it could be `&`, then pressing `&&` would open the crafting menu and in that the choose crafter menu.

I consider `uimenu`'s implementation temporary, but the public interface might as well be final.

### Testing

1. Have a friendly NPC who knows next to nothing.
2. Through the debug menu, obtain all recipes and get all skills to 10 for yourself.
3. Open the crafting menu `&` and switch between you and the NPC `c` (new keybinding) as you browse the crafting menu.
4. Observe that NPC cannot craft recipes that they shouldn't be able to (cannot craft recipe whose `floor(recipe.primary_skill_requirement * 0.8) > guy.skill`). If you press enter, it says `Crafter cannot craft that!`.
5. Observe that recipes have `Known by You` (this is a bugfix).
6. Select a recipe and start crafting.
7. Observe that the selected crafter is crafting the recipe (even if somebody else opened the crafting menu).



### Additional context

Allies cannot use items from each other's inventories. #68182 Somebody please implement it :)

Now, switching between characters will shuffle recipes dramatically, as some become (un)craftable, it would be solved by implementing the aforementioned.

You can tell NPC to craft a crossbow trap out of their crossbow. I hope this is in line. This PR didn't change that behaviour but I noticed it thanks to it.

Switching between characters filters by `filterstring` again (notably slow with prompts like `d:close to`), but it shouldn't as the list of recipes is the same, since crafting_group couldn't have changed.
